onnxruntime/contrib_ops/cpu/reverse.cc

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "reverse.h"
+#include <utility>
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/framework/utils.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Reverse,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Reverse);
+
+Status handle_scalar_tensor(const Tensor* input_tensor, Tensor* output_tensor, const MLDataType& dtype) {
+  // for scalar tensors, "reversing" is just copying the input buffer to the output buffer
+
+  // string scalar
+  if (dtype == DataTypeImpl::GetType<std::string>()) {
+    const std::string* src = input_tensor->template Data<std::string>();
+    std::string* dst = output_tensor->template MutableData<std::string>();
+    if (src != dst)
+      std::copy(src, src + 1, dst);
+    return Status::OK();
+  }
+
+  // MLFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<MLFloat16>()) {
+    const auto* src = input_tensor->Data<MLFloat16>();
+    auto* dst = output_tensor->template MutableData<MLFloat16>();
+    dst[0] = MLFloat16(math::floatToHalf(math::halfToFloat(src[0].val)));
+    return Status::OK();
+  }
+
+  // BFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<BFloat16>()) {
+    const auto* src = input_tensor->Data<BFloat16>();
+    auto* dst = output_tensor->template MutableData<BFloat16>();
+    dst[0] = BFloat16(src[0].ToFloat());
+    return Status::OK();
+  }
+
+  // non-string and non-float16 scalars - just copy the raw bytes
+  const void* src = input_tensor->DataRaw(dtype);
+  void* dst = output_tensor->MutableDataRaw(dtype);
+  if (src != dst)
+    memcpy(dst, src, 1 * dtype->Size());
+
+  return Status::OK();
+}
+
+template <int rank>
+void ReverseImplMLFloat16Type(const OpKernelContext* ctx, const Tensor* input_tensor, Tensor* output_tensor, const TensorShape& shape, const std::vector<int64_t>& reverse_axes) {
+  AllocatorPtr allocator;
+  ctx->GetTempSpaceAllocator(&allocator);
+  ORT_ENFORCE(ctx->GetTempSpaceAllocator(&allocator).IsOK(), "ORT needs couldn't get access to intermediate memory allocator to process float16 data in Reverse operator");
+  ORT_ENFORCE(allocator != nullptr, "ORT needs couldn't get access to intermediate memory allocator to process float16 data in Reverse operator");
+
```

Comment:
2 out of 3 of these lines seem unnecessary. 
I think the 'ORT' is implied and the error message could be simpler and just say that getting the temporary allocator failed. That failure isn't due to us processing a particular data type so that information could be excluded here.

Reply:
I agree. Removed line 2.